test/cases/parsing/issue-4940/index.js

```diff
@@ -0,0 +1,11 @@
+it("should create dependency when require is called with 'new' (object export)", function() {
+	const result = new require("./object-export");
+	result.foo.should.equal("bar");
+});
+
+it("should create dependency when require is called with 'new' (non-object export)", function() {
+	const sideEffect = require("./sideEffect");
+	const result = new require("./non-object-export");
+	result.should.instanceof(__webpack_require__);
+	sideEffect.foo.should.equal("bar");
+});
```

Comment:
Also add a test case like this:

``` js
define("local-module", function() { return 1; })

const result = new require("local-module");
result.should.instanceof(__webpack_require__);
```

Reply:
@sokra I'm not very familiar with this. Are you referring to [this functionality](https://github.com/webpack/docs/wiki/amd#define)? `define` can be freely called in a webpack bundle?
@ooflorent in that case the function would be the module factory and would be equivalent to
```js
// a.js
module.exports = 1;
```
I think. Then the test holds.

Reply:
@sokra Indeed, 
```js
define("local-module", function() { return 1; })
const result = new require("local-module");
result.should.instanceof(__webpack_require__);
```
this fails because `result === 1`.

Looks like it isn't as easy as putting `new` in front since those local modules are instantiated with `.call`. I think we can get around that by doing

```js
localModule.exports = (Function.prototype.bind.call(function () {
  return 1;
}, /* arguments here */)()
```
instead of 
```js
localModule.exports = function () {
  return 1;
}.call( /* arguments here */)
```
then, if we want, we can interpollate `new` like 
```js
localModule.exports = new (Function.prototype.bind.call(function () {
  return 1;
}, /* arguments here */)()
```
But even then it still wouldn't be `instanceof __webpack_require__`.
(not my idea, stolen from [SO](https://stackoverflow.com/questions/1606797/use-of-apply-with-new-operator-is-this-possible))

I'm trying it out locally here, but this feels a bit like a rabbit hole. Are you sure this "correctness" is worth the effort? The only case this would mess up someone's code is when they rely on `new require('local-module')` to return a nonsense object instead of a primitive. I'm gonna argue such code doesn't even exist, and if someone writes it they deserve weird errors 🙂

Reply:
> Are you sure this "correctness" is worth the effort?

Absolutely not! I wanted to point out that webpack's behavior will differ from node.

Reply:
A trivial implementation of `require` is:

```js
function _load(id) {
  const module = …
  return module.exports
}

function require(id) {
  return _load(id)
}
```

Since `require` returns `exports` it behaves like _function_ constructors:

```js
function F() { return 1 }
console.log(new F, (new F) instanceof F) // F {}  true

function G() { return {a: 1} }
console.log(new G, (new G) instanceof G) // {a: 1}  false

function H() { return function() { this.a = 2} }
console.log(new H, (new H) instanceof H) // Function  false
console.log(new (new H)) // {a: 2}
```

I don't think we could easily implement this at compile time.

Reply:
**tl;dr:** `new require(a)` does nothing more than `require(a)` unless `a` exports a non-object value.

Reply:
Ok the instanceof thing is too much. We can generate this code:

``` js
new require("local-module")
new (function() { return localModule; })()
```

So it's at least correctly a object

Reply:
@sokra Ok, just so we're on the same page, let's go over what needs to happen in practice:
All these function defines are instanciated with `.call` or `.apply` like [here](https://github.com/webpack/webpack/blob/master/lib/dependencies/AMDDefineDependency.js#L60). This means we can't simply add `new` in front of it based on a flag. This means we 

1. either turn these `fn.call(...)` into `(Function.prototype.bind.call(fn, ...))()` so that we can piut a `new` in front of it conditionally
2. either we add a new flag on top of existing `l`, `a`, `o` and `f`. Maybe `n`. Then define all new templates based on the `new (function() { return localModule; })()` pattern. Ofcourse they will have a different `this` binding since we can't control it when calling `new`. So these functions will essentially execute totally differently than the existing `define`s.

Unless I'm missing something, 2. changes the semantics of local modules. So we'll have to go with solution 1. Are you sure this is worth the extra code size? 
Also, which version of javascript do the resulting bundles target? 2015 or older?

Reply:
Also, I just realized, what if somebody does both
```js
const x = require("local-module")
```
and
```js
const y = new require("local-module")
```
in the same code base?

Reply:
I had another idea overnight. Will experiment further.